[93] feat: first iteration of spec

Savio629 opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/nest/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: https://github.com/SamagraX-Stencil/stencil/issues/93


## What is the new behavior?


## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

techsavvyash commented on actions/list.action.ts:
```diff
@@ -19,18 +19,18 @@ export class ListAction extends AbstractAction {
   }
   private async loadRegistry(): Promise<void>  {
-        const stencilFilePath = join(process.cwd(), '.stencil');
+        const stencilFilePath = join(process.cwd(), 'spec.yaml');
```
don't hardcode the name, take this as input from the user, the user should be able to specify the path where the spec file is located.

techsavvyash commented on actions/new.action.ts:
```diff
@@ -213,13 +224,13 @@ const askForMissingInformation = async (inputs: Input[], options: Input[]) => {
 
   const prismaInput = getPrismaInput(options);
-  if (!prismaInput!.value) {
+  if (prismaInput && prismaInput.value === undefined) {
```
this condition has been reused mutitple times, can be abstract this out as a function call and optimise the code using loops and function calls if possible?

techsavvyash commented on actions/new.action.ts:
```diff
@@ -674,37 +685,39 @@ const copyEnvFile = async (dir: string, envExample: string, envFile: string) =>
-  shouldInitializeLogging: boolean,
-  shouldInitializeFileUpload: boolean
+  projectDirectory: string,
+  spec: {
+    projectName: string,
+    prismaSetup: boolean,
+    userServiceSetup: boolean,
+    monitoringSetup: boolean,
+    monitoringServicesSetup: boolean,
+    temporalSetup: boolean,
+    loggingSetup: boolean,
+    fileUploadSetup: boolean,
+    packageManager: string
+  }
```
define a type for the spec as well.

techsavvyash commented on actions/new.action.ts:
```diff
@@ -674,37 +685,39 @@ const copyEnvFile = async (dir: string, envExample: string, envFile: string) =>
-  ].filter(info => info !== '').join('\n');
+  const { projectName, packageManager, ...setupInfo } = spec;
 
+  const filePath = join(process.cwd(), projectDirectory, 'spec.yaml');
```
don't hardcode file name.

techsavvyash commented on actions/new.action.ts:
```diff
@@ -674,37 +685,39 @@ const copyEnvFile = async (dir: string, envExample: string, envFile: string) =>
   try {
     await fs.promises.access(filePath, fs.constants.F_OK);
-    console.log('.stencil file already exists');
+    console.log('spec.yaml file already exists');
```
either convert this to a proper logger or remove.

Savio629 replied:
What do you mean by proper logger?
Like using chalk and passing Message parameter?

techsavvyash replied:
yes

techsavvyash commented on actions/new.action.ts:
```diff
@@ -674,37 +685,39 @@ const copyEnvFile = async (dir: string, envExample: string, envFile: string) =>
+      ...setupInfo
+    });
+    await fs.promises.writeFile(filePath, yamlContent);
+    console.log('spec.yaml file created');
```
either convert this to a proper logger or remove.

techsavvyash commented on actions/spec.action.ts:
```diff
@@ -0,0 +1,79 @@
+            console.log(chalk.green('Specification file'));
+            console.log(stencilFileContent);
+            const spec = yaml.load(stencilFileContent) as {
+              projectName: string;
+              packageManager: string;
+              prismaSetup: boolean;
+              userServiceSetup: boolean;
+              monitoringSetup: boolean;
+              monitoringServicesSetup: boolean;
+              temporalSetup: boolean;
+              loggingSetup: boolean;
+              fileUploadSetup: boolean;
+          };          
```
define this as a type and then typecast.

techsavvyash commented on actions/spec.action.ts:
```diff
@@ -0,0 +1,79 @@
+              fileUploadSetup: boolean;
+          };          
+
+          let command = `stencil new ${spec.projectName}`;
+          command += ` --ps ${spec.prismaSetup ? 'yes' : 'no'}`;
+          command += ` --us ${spec.userServiceSetup ? 'yes' : 'no'}`;
+          command += ` --mo ${spec.monitoringSetup ? 'yes' : 'no'}`;
+          command += ` --ms ${spec.monitoringServicesSetup ? 'yes' : 'no'}`;
+          command += ` --te ${spec.temporalSetup ? 'yes' : 'no'}`;
+          command += ` --lg ${spec.loggingSetup ? 'yes' : 'no'}`;
+          command += ` --fu ${spec.fileUploadSetup ? 'yes' : 'no'}`;          
+            command += ` --package-manager ${spec.packageManager}`;
```
use loops
```suggestion
Object.keys(spec).forEach(key => {
   command += // however you'd build the logic.
})

techsavvyash commented on actions/spec.action.ts:
```diff
@@ -0,0 +1,79 @@
+          command += ` --lg ${spec.loggingSetup ? 'yes' : 'no'}`;
+          command += ` --fu ${spec.fileUploadSetup ? 'yes' : 'no'}`;          
+            command += ` --package-manager ${spec.packageManager}`;
+            console.log(command);
```
either use a logger or remove this.

techsavvyash commented on commands/new.command.ts:
```diff
@@ -32,38 +33,61 @@ export class NewCommand extends AbstractCommand {
-        'If you want prisma setup in the project',
-      )
+      .option('--ps, --prisma [prisma]', 'If you want prisma setup in the project', (value) => {
+        // if (value === 'true') return true;
+        // if (value === 'false') return false;
```
remove redundant comments.

techsavvyash commented on commands/new.command.ts:
```diff
@@ -32,38 +33,61 @@ export class NewCommand extends AbstractCommand {
-        'If you user service setup in the project',
-      )
+        'If you user service setup in the project', (value) => {
+          // if (value === 'true') return true;
+          // if (value === 'false') return false;
```
remove redundant comments.

techsavvyash commented on commands/new.command.ts:
```diff
@@ -32,38 +33,61 @@ export class NewCommand extends AbstractCommand {
-        'If you want to have fixtures in the project',
-      )
+        'If you want to have fixtures in the project',(value) => {
+          // if (value === 'true') return true;
+          // if (value === 'false') return false;
```
remove redundant comments.

techsavvyash commented on commands/new.command.ts:
```diff
@@ -32,38 +33,61 @@ export class NewCommand extends AbstractCommand {
-        'If you want to have monitoring setup in the project',
-      )
+        'If you want to have monitoring setup in the project',(value) => {
+          if (value === 'true') return true;
+          if (value === 'false') return false;
+          return value;
```
```suggestion
return value === "true"
```

techsavvyash commented on commands/new.command.ts:
```diff
@@ -32,38 +33,61 @@ export class NewCommand extends AbstractCommand {
-        'If you want to have monitoring-service setup in the project',
-      )
+        'If you want to have monitoring-service setup in the project',(value) => {
+          if (value === 'true') return true;
+          if (value === 'false') return false;
+          return value;
```
```suggestion
return value === "true"
```

techsavvyash commented on commands/new.command.ts:
```diff
@@ -32,38 +33,61 @@ export class NewCommand extends AbstractCommand {
-        'If you want to have temporal setup in the project',
-      )
+        'If you want to have temporal setup in the project',(value) => {
+          if (value === 'true') return true;
+          if (value === 'false') return false;
+          return value;
```
```suggestion
return value === "true"
```

techsavvyash commented on commands/new.command.ts:
```diff
@@ -32,38 +33,61 @@ export class NewCommand extends AbstractCommand {
-        'If you want to have logging setup in the project',
-      )
+        'If you want to have logging setup in the project',(value) => {
+          if (value === 'true') return true;
+          if (value === 'false') return false;
+          return value;
```
```suggestion
return value === "true"
```

techsavvyash commented on commands/new.command.ts:
```diff
@@ -32,38 +33,61 @@ export class NewCommand extends AbstractCommand {
-        'If you want to have fileUpload setup in the project',
-      )
+        'If you want to have fileUpload setup in the project',(value) => {
+          if (value === 'true') return true;
+          if (value === 'false') return false;
+          return value;
```
```suggestion
return value === "true"
```

techsavvyash commented on commands/new.command.ts:
```diff
@@ -116,8 +137,16 @@ export class NewCommand extends AbstractCommand {
         const inputs: Input[] = [];
         inputs.push({ name: 'name', value: name });
-
+        console.log(options);
```
remove

Savio629 commented:
@techsavvyash 
The generated spec should store the below pattern right?
```
stencil: 0.0.1

info:
  properties:
    project-name: "Final"
    package-manager: "npm"

tooling: []

endpoints:
```

instead of the previous registry?

```
projectName: Final
packageManager: npm
prismaSetup: false
userServiceSetup: false
monitoringSetup: false
monitoringServicesSetup: false
temporalSetup: false
loggingSetup: false
fileUploadSetup: false
```

so how should stencil ls look?
Should it log the first file completely or log in the second structure?

techsavvyash commented:
> @techsavvyash 
> 
> The generated spec should store the below pattern right?
> 
> ```
> 
> stencil: 0.0.1
> 
> 
> 
> info:
> 
>   properties:
> 
>     project-name: "Final"
> 
>     package-manager: "npm"
> 
> 
> 
> tooling: []
> 
> 
> 
> endpoints:
> 
> ```
> 
> 
> 
> instead of the previous registry?
> 
> 
> 
> ```
> 
> projectName: Final
> 
> packageManager: npm
> 
> prismaSetup: false
> 
> userServiceSetup: false
> 
> monitoringSetup: false
> 
> monitoringServicesSetup: false
> 
> temporalSetup: false
> 
> loggingSetup: false
> 
> fileUploadSetup: false
> 
> ```
> 
> 
> 
> so how should stencil ls look?
> 
> Should it log the first file completely or log in the second structure? 

Yes

Savio629 commented:
What should it log then when stencil ls is used?
First or second?

Savio629 commented:
#### Features or fix implemented till now in this pr

- implemented `stencil spec <path of spec file>`
- updated `stencil ls` to load new spec structure
- added `"specFile": "spec.yaml"` in nest-cli.json after creation of registry

added loops where necessary and proper logger statements

 stencil spec <link of the spec file> should also work?

Savio629 commented:
README.md for Spec Command


## Spec Command

```
stencil spec [filePath]
stencil sp [filePath]
```
**Description**

Bootstraps the project based on given specification file.

**Arguments**

| Argument  |  Description |
|-----------|--------------|
|  `[filePath]`	 | The specification file along with path |

## Structure of spec.yaml
```
stencil: 0.0.1

info:
  properties:
    project-name: "ProjectName"
    package-manager: "npm" 
    
tooling: [logging]

endpoints:
```

*Supported Package Managers are npm, yarn, pnpm and bun.*
*Supported toolings are prisma, userService, temporal, monitoring, monitoringService, logging, fileUpload*

techsavvyash commented on Spec-readme.md:
```diff
@@ -0,0 +1,38 @@
+    
+tooling: [logging]
+
+endpoints:
```
What is going to be the content of this key? Can't see that detailed out anywhere.

Savio629 replied:
Once the crud api get's merge I'll update the endpoints structure.
Also wanted to know how should it work?
Can you give some suggestions here?

techsavvyash commented on Spec-readme.md:
```diff
@@ -0,0 +1,38 @@
+
+*Supported Package Managers are npm, yarn, pnpm and bun.*
+
+*Supported toolings are prisma, userService, temporal, monitoring, monitoringService, logging, fileUpload*
```
Please explain what each one does and when it can be used.

Savio629 replied:
Okay

techsavvyash commented on actions/list.action.ts:
```diff
@@ -1,39 +1,52 @@
+
+  constructor() {
+    super();
+    this.specFilePath = 'spec.yaml'; // Default spec file path if not found in nest-cli.json
```
Why is this path hardcoded?

Savio629 replied:
default spec.yaml path if no path is found inside nest-cli.json

Savio629 replied:
also the spec file will be generated automatically by the compiler when the project is being bootstrap.

techsavvyash commented on actions/list.action.ts:
```diff
@@ -1,39 +1,52 @@
+      const stencilFileContent = await fs.promises.readFile(stencilFilePath, 'utf-8');
 
+      if (stencilFileContent.trim() === '') {
+        console.info(chalk.yellow(MESSAGES.NO_SPEC_FOUND));
```
Since we are not erroring out what is the behaviour incase the spec file is empty? Can't seem to find it in the docs.

Savio629 replied:
Right!
Will update that test case and add the condition under stencil spec command

Savio629 replied:
if spec.yaml is empty I was doing console.info "No specification file found". Should I change it to "Spec file empty" ?

techsavvyash commented on actions/new.action.ts:
```diff
@@ -674,33 +654,63 @@ const copyEnvFile = async (dir: string, envExample: string, envFile: string) =>
-  ].filter(info => info !== '').join('\n');
+  const { projectName, packageManager, ...setupInfo } = spec;
 
+  const filePath = join(process.cwd(), projectDirectory, 'spec.yaml');
```
Why is the file path hardcoded?

Savio629 replied:
The spec.yaml file will be generated everytime inside the root folder of the project right?
if `stencil spec temp/spec.yaml` is being run by the user and the spec.yaml has projectName `final` then the spec.yaml will be generated inside of root of `final` project folder, right?

techsavvyash commented on actions/new.action.ts:
```diff
@@ -674,33 +654,63 @@ const copyEnvFile = async (dir: string, envExample: string, envFile: string) =>
+    if (setupInfo.fileUploadSetup) tooling.push('fileUpload');
+
+    const yamlContent = yaml.dump({
+      stencil: '0.0.1',
```
Why is the version hardcoded? What if we want to update the version or specify an older version of stencil to use?

Savio629 replied:
Wanted to know from where should I retrieve this?
The stencil package version, right? (This is where the spec.yaml file is being created when `stencil new` command is run)

techsavvyash commented on actions/spec.action.ts:
```diff
@@ -0,0 +1,112 @@
+      const stencilFileContent = await fs.promises.readFile(stencilFilePath, 'utf-8');
+      
+      if (stencilFileContent.trim() === '') {
+        console.info(chalk.yellow(MESSAGES.NO_SPEC_FOUND));
```
Throw here

techsavvyash commented on test/e2e-test/spec.e2e-spec.ts:
```diff
@@ -0,0 +1,44 @@
+
+  it('should run the Stencil CLI with the "spec.yaml" file and check for directory structure', (done) => {
+    writeFileSync(specFilePath, specFileContent);
+    exec(`npx stencil spec ${specFilePath}`, (error, stdout, stderr) => {
```
Why are we running a command via npx to test this very package?? I don't think npx will target and use this repo. How is this test case even passing on your local? Or have you pushed without even running this?

techsavvyash commented on test/e2e-test/spec.e2e-spec.ts:
Can we do a better job at testing and have more and *CORRECT* tests?

techsavvyash commented on test/e2e-test/spec.e2e-spec.ts:
```diff
@@ -0,0 +1,109 @@
+
+describe('Stencil cli e2e Test - SPEC command', () => {
+  const specFilePath = join('spec.yaml');
+  const specFileContent = `stencil: 0.0.1
+
+info:
+  properties:
+    project-name: "final"
+    package-manager: "npm"
+
+tooling: [temporal]
+
+endpoints:
+`;
```
read from a file.

techsavvyash commented on test/e2e-test/spec.e2e-spec.ts:
```diff
@@ -0,0 +1,109 @@
+  });
+
+  it('should log an error if the package-manager is missing in the spec file', (done) => {
+    const invalidSpecFileContent = `stencil: 0.0.1
+  info:
+    properties:
+      project-name: "final"
+  tooling: [temporal]`;
```
read from a file

techsavvyash commented on actions/spec.action.ts:
```diff
@@ -0,0 +1,142 @@
+        const dockerCommand = `stencill docker ${dockerServices.join(' ')}`;
+        console.info(chalk.green(`Initializing Docker services: ${dockerServices.join(', ')}`));
+        process.chdir(projectDir);
+        // await this.runCommand('npm link @samagra-x/schematics @samagra-x/stencil-stencil'); # For testing
```
remove this comment

techsavvyash commented on actions/new.action.ts:
```diff
@@ -129,11 +116,11 @@ export class NewAction extends AbstractAction {
-      projectDirectory,
-      shouldInitializeLogging as boolean,
-    );
+    // await createLogging(
+    //   isDryRunEnabled as boolean,
+    //   projectDirectory,
+    //   shouldInitializeLogging as boolean,
+    // );
```
remove
